main.go

```diff
@@ -0,0 +1,190 @@
+package k8s
+
+import (
+	"strings"
+
+	"github.com/alibaba/sentinel-golang/logging"
+	"github.com/alibaba/sentinel-golang/util"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	datasourcev1 "github.com/sentinel-group/sentinel-go-datasource-k8s-crd/api/v1"
+	"github.com/sentinel-group/sentinel-go-datasource-k8s-crd/controllers"
+	// +kubebuilder:scaffold:imports
+)
+
+var (
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+func init() {
+	_ = clientgoscheme.AddToScheme(scheme)
+
+	_ = datasourcev1.AddToScheme(scheme)
+	// +kubebuilder:scaffold:scheme
+}
+
+type CRDType int32
+
+const (
+	FlowRulesCRD CRDType = iota
+	IsolationRulesCRD
+	CircuitBreakerRulesCRD
+	HotspotRulesCRD
+	SystemRulesCRD
+)
+
+func (c CRDType) String() string {
+	switch c {
+	case FlowRulesCRD:
+		return "FlowRulesCRD"
+	case CircuitBreakerRulesCRD:
+		return "CircuitBreakerRulesCRD"
+	case HotspotRulesCRD:
+		return "HotspotRulesCRD"
+	case SystemRulesCRD:
+		return "SystemRulesCRD"
+	default:
+		return "Undefined"
+	}
+}
+
+type K8SDataSource struct {
+	crdManager  ctrl.Manager
+	controllers map[CRDType]reconcile.Reconciler
+	stopChan    chan struct{}
+}
+
+func NewK8SDataSource() (*K8SDataSource, error) {
+	ctrl.SetLogger(&k8SLogger{
+		l:             logging.GetGlobalLogger(),
+		level:         logging.GetGlobalLoggerLevel(),
+		names:         make([]string, 0),
+		keysAndValues: make([]interface{}, 0),
+	})
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme: scheme,
+		// disable metric server
+		MetricsBindAddress:     "0",
+		HealthProbeBindAddress: "0",
+		LeaderElection:         false,
+	})
+	if err != nil {
+		setupLog.Error(err, "unable to start manager")
+		return nil, err
+	}
+	k := &K8SDataSource{
+		crdManager:  mgr,
+		controllers: make(map[CRDType]reconcile.Reconciler, 4),
+		stopChan:    make(chan struct{}),
+	}
+	return k, nil
+}
+
+func (k *K8SDataSource) RegisterController(crd CRDType, crName string) error {
```

Comment:
One more `namespace` filter is needed so that we could distinguish different configs from different namespaces.
Also we may need some comments here.

Reply:
We might need DataSource level namespace .

Reply:
Agree.